### PR TITLE
Direct socket-based communication

### PR DESCRIPTION
**Pull request by darrylmasson**
Rather than having everything poke the database once per second to see if there are new commands, it is much faster to have things communicate directly. This PR rewrites the command subsystem to enable this. All Monitors now have a Listener, which picks an open port and listens for incoming commands. The hypervisor has a Dispatcher, which acts as an intermediary between things to facilitate delayed commands. All commands are routed via the dispatcher, which maintains an internal queue. Assuming the rate of internal message-passing doesn't exceed 10s of messages per second, no bottlenecking is expected.

### CODE REVIEW

File: Doberman/utils.py [hunk 1]
```diff
@@ -287,3 +287,59 @@ def sensible_sig_figs(reading, lowlim, upplim, defaultsigfigs=3):
...
+        if len(self._buf) == 0:
+            self._buf.append(obj)
+        elif len(self._buf) == 1:
+            if self._buf[0]['time'] >= obj['time']:
+                self._buf.insert(0, obj)
+            else:
+                self._buf.append(obj)
+        else:
+            idx = len(self._buf)//2
+            for i in itertools.count(2):
```
Comment (jarongrigat) [inline Doberman/utils.py:317]: `import itertools`

File: Doberman/hypervisor.py [hunk 2]
```diff
@@ -108,31 +115,74 @@ def start_pipeline(self, pipeline: str) -> int:
...
+                    if doc['to'] == 'hypervisor':
+                        self.process_command(doc['command'])
+                        continue
+                    hn, p = self.db.get_listener_address(doc['to'])
+                    if self.db.get_sensor_setting(name=doc['to'], field='status') != 'online':
+                        self.logger.warning(f'Can\'t send command "{doc["command"]}" to {doc["to"]} '
+                                f' because it isn\'t online')
+                        continue
+                    self.logger.debug(f'Sending "{doc["command"]}" to {doc["to"]}')
+                    with socket.create_connection((hn, p), timeout=0.1) as sock:
```
Comment (jarongrigat) [inline Doberman/hypervisor.py:147]: `import socket`

File: Doberman/hypervisor.py [hunk 3]
```diff
@@ -108,31 +115,74 @@ def start_pipeline(self, pipeline: str) -> int:
...
+            except Exception as e:
+                self.logger.info(f'Dispatcher caught a {type(e)}: {e}')
+        self.logger.debug('Dispatcher shutting down')
+
+    def process_command(self, command) -> None:
+        self.logger.debug(f'Processing {command}')
+        if command[0] == '{':
+            # a json document, this is for the dispatcher
+            with self.cv:
+                self.dispatch_queue.add(json.loads(command))
```
Comment (jarongrigat) [inline Doberman/hypervisor.py:158]: `import json`

**Review by jarongrigat — CHANGES_REQUESTED** [review]

**Review by jarongrigat — APPROVED** [review]

**Comment by darrylmasson** [comment]
Ok, I think we're ready to merge

**Pull request merged**